### ISSUE TITLE
masked_scatter change elements even the the self and source tensor are of same values

### ISSUE THREAD

**Issue by zxydi1992** [pytorch/pytorch#28598]
## 🐛 Bug

Tensor.masked_scatter() change elements even the the self and source tensor are of same values

## To Reproduce

1. run the script below: 
#!/usr/bin/env python
import torch
import numpy as np

t1 = torch.rand(16, 100)
t2 = t1.clone()

m = np.zeros((1, 100), dtype=np.int64)
m[0, :10] = 1
m = torch.tensor(m, dtype=torch.bool)

print(t1)

print(t2)

print(t1.masked_scatter(m, t2))

## Expected behavior

It holds that:t1.masked_scatter(m, t2) == t1 and t1.masked_scatter(m, t2) == t2, 
since t1 has the same value as t2.

## Environment

Collecting environment information...
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.13.6
GCC version: Could not collect
CMake version: version 3.12.2

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] idx2numpy==1.2.2
[pip] msgpack-numpy==0.4.3.2
[pip] numpy==1.15.0
[pip] numpydoc==0.9.1
[pip] pytorch-ignite==0.2.0
[pip] pytorch-pretrained-bert==0.6.2
[pip] pytorch-transformers==1.1.0
[pip] torch==1.2.0
[pip] torch-cluster==1.4.4
[pip] torch-geometric==1.3.1
[pip] torch-scatter==1.3.1
[pip] torch-sparse==0.4.0
[pip] torch-spline-conv==1.1.0
[pip] torchfile==0.1.0
[pip] torchtext==0.4.0
[pip] torchvision==0.4.0a0+6b959ee
[conda] _tflow_select             2.3.0                       mkl
[conda] blas                      1.0                         mkl
[conda] mkl                       2018.0.3                      1
[conda] mkl_fft                   1.0.6            py36hb8a8100_0
[conda] mkl_random                1.0.1            py36h5d10147_1
[conda] pytorch                   1.2.0                   py3.6_0    pytorch
[conda] pytorch-ignite            0.2.0                    pypi_0    pypi
[conda] pytorch-pretrained-bert   0.6.2                    pypi_0    pypi
[conda] pytorch-transformers      1.1.0                    pypi_0    pypi
[conda] tensorflow                1.9.0           mkl_py36h4fc8490_1
[conda] tensorflow-base           1.9.0           mkl_py36h70e0e9a_0
[conda] tensorflow-mkl            1.9.0                h067cd57_0
[conda] torch-cluster             1.4.4                    pypi_0    pypi
[conda] torch-geometric           1.3.1                    pypi_0    pypi
[conda] torch-scatter             1.3.1                    pypi_0    pypi
[conda] torch-sparse              0.4.0                    pypi_0    pypi
[conda] torch-spline-conv         1.1.0                    pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchtext                 0.4.0                    pypi_0    pypi
[conda] torchvision               0.2.0                    pypi_0    pypi

## Additional context

It also occured on a Ubuntu 16.04 GPU server with pytorch 1.3.0.

**Comment by zhangguanheng66** [comment]
I have a feeling that it could be a promotion issue. Similar issue https://github.com/pytorch/pytorch/issues/28508
https://github.com/pytorch/pytorch/issues/28502

**Comment by hi-ogawa** [comment]
Hi @zxydi1992, I think your expectation about `masked_scatter` is not correct here and it does not (necessarily) satisfy the equation you showed `t.masked_scatter(mask, t) == t`.
If carefully read documentation (of [masked_scatter_](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.masked_scatter_)), it says:

> The mask operates on the `self` tensor, not on the given `source` tensor.

which means, for example, it does something like this:

```python
import torch
self = torch.tensor([1, 2, 3, 4, 5, 6, 7])
mask = torch.tensor([1, 0, 0, 1, 0, 1, 1], dtype=torch.bool)
src = torch.tensor([100, 200, 300, 400]) # src entry is picked up just "from left to right"

self.masked_scatter(mask, src)
# => tensor([100,   2,   3, 200,   5, 300, 400])
```

So, I believe, it's only required that:

- `mask.sum() <= src.numel()` (i.e. number of true entries in `mask` <= number of element of `src`)
-  `mask.shape() == self.shape()` (or `mask` is broadcast-able to `self`)

For the reference, it's implemented like this (for cpu):

https://github.com/pytorch/pytorch/blob/5da932ad728ed61f3fe09ccb7dfa8a4795a45efb/aten/src/TH/generic/THTensorEvenMoreMath.cpp#L241-L267

Anyway, I hope my example helps you.

**Comment by zhangguanheng66** [comment]
@zxydi1992 Could you check the master branch and report it here. Feel free to reopen the issue if you still have questions.

**Comment by zxydi1992** [comment]
@hi-ogawa Thank you for the suggestion. I believe we should have some more documents on the behavior of masked_scatter. 

Btw, @zhangguanheng66,  do we have a way to do a masked copy on two tensors in the following manner?

x = some tensor(d1, d2, d3)
y = some tensor(d1, d2, d3)
m = some boolean tensor (d1, d2, d3)

z = x.masked_copy(m, y)
such that 

z[i][j][k] == x[i][j][k] if m[i][j][k] = False and
z[i][j][k] == y[i][j][k] if m[i][j][k] = True

As suggested by @hi-ogawa , the current masked_scatter method is not designed for this perpose. Plus, It could be better if the above operation support broadcasting.

**Comment by zhangguanheng66** [comment]
you may consider `torch.where`.
